### DISCUSSION TITLE
Screen CSS continues to apply after pop

### DISCUSSION THREAD

**Discussion by ndevenish** [Textualize/textual#2911] [General]
Screen CSS taints the CSS for the whole app, even after the screen is removed from the screen stack. This doesn't seem like it should happen.

It looks like support for `Screen.CSS` was added in https://github.com/Textualize/textual/pull/2744, which did identify this [footgun](https://github.com/Textualize/textual/pull/2744#discussion_r1226787029). Applying CSS that continues to apply after the screen no longer exists definitely feels like a bug, because it doesn't give any different effects from Application-wide applied CSS.

I encountered this when I wanted to add a loading screen so promoted my topmost widget to a screen.

```python
from textual.app import App
from textual.screen import Screen
from textual.widgets import Static


class Over(Screen):
    CSS = "Static { background: green; }"

    def compose(self):
        yield Static(f"Message on top screen")
        self.set_interval(1, lambda: self.screen.app.pop_screen())


class Under(Screen):
    def compose(self):
        yield Static("Message on bottom screen")


class ScreenApp(App):
    def on_mount(self):
        self.push_screen(Under())
        self.push_screen(Over())


if __name__ == "__main__":
    ScreenApp().run()
```
Textual 0.29.0. The `Static` on screen `Under` continues to have a green background, even when the other screen is removed. I'd expect it not to.

**Comment by ndevenish** [comment]
Easily worked around btw by scoping all of the second Screen CSS to screen-specific classes (or continuing to use a top-level-widget that contains the CSS)

**Comment by rodrigogiraoserrao** [comment]
Hey @ndevenish, we have something in the works that will make `Screen.CSS` behave in a slightly more controlled manner.
Until then, being more strict with the selectors used may be the way to go 👍

**Comment by willmcgugan** [comment]
This is the expected behaviour at the moment. The solution, as you discovered, is to scope the CSS to your screen.

I guess it was just a test, but you would be unlikely to style a `Static` so generally, as you may end up styling more that desired.

The advice may change when we implement "nested" css.

**Comment by ndevenish** [comment]
The actual literal implementation to style the Static that triggered this case wasn't far off; a loading screen:
```python
class LoadingScreen(Screen):
    CSS = """
    Screen {
        align: center middle;
        layers: bottom top;
    }
    Static {
        content-align: center middle;
        text-align: center;
        offset: 0 -200%;
        layer: top;
    }
    LoadingIndicator { layer: bottom; }
    """

    def compose(self):
        yield Static("Message")
        yield LoadingIndicator()
```
(update/removal logic stripped here). Unfortunately an offset isn't as obvious as a background and caused lots of weird behaviour.

> This is the expected behaviour at the moment. The solution, as you discovered, is to scope the CSS to your screen.

What does having `Screen.CSS` enable that the `App.CSS` property doesn't already do, with less potential for confusion? An intention to allow `Screen` to be the top-level object without an `App`? Or is this just placeholder behaviour until the cascading- part of the style sheets applies to it?
